server/src/main/java/org/opensearch/common/cache/store/EhCacheDiskCache.java

```diff
@@ -0,0 +1,543 @@
+
+    // Ehcache disk write maximum threads for its pool
+    public final Setting<Integer> DISK_WRITE_MAXIMUM_THREADS;
+
+    // Not be to confused with number of disk segments, this is different. Defines
+    // distinct write queues created for disk store where a group of segments share a write queue. This is
+    // implemented with ehcache using a partitioned thread pool exectutor By default all segments share a single write
+    // queue ie write concurrency is 1. Check OffHeapDiskStoreConfiguration and DiskWriteThreadPool.
+    public final Setting<Integer> DISK_WRITE_CONCURRENCY;
+
+    // Defines how many segments the disk cache is separated into. Higher number achieves greater concurrency but
+    // will hold that many file pointers. Default is 16.
+    public final Setting<Integer> DISK_SEGMENTS;
+
+    private final StoreAwareCacheEventListener<K, V> eventListener;
+
+    /**
+     * Used in computeIfAbsent to synchronize loading of a given key. This is needed as ehcache doesn't provide a
+     * computeIfAbsent method.
+     */
+    Map<K, CompletableFuture<Tuple<K, V>>> completableFutureMap = new ConcurrentHashMap<>();
+
+    private EhCacheDiskCache(Builder<K, V> builder) {
+        this.keyType = Objects.requireNonNull(builder.keyType, "Key type shouldn't be null");
+        this.valueType = Objects.requireNonNull(builder.valueType, "Value type shouldn't be null");
+        this.expireAfterAccess = Objects.requireNonNull(builder.getExpireAfterAcess(), "ExpireAfterAccess value shouldn't " + "be null");
+        this.maxWeightInBytes = builder.getMaxWeightInBytes();
+        if (this.maxWeightInBytes <= MINIMUM_MAX_SIZE_IN_BYTES) {
+            throw new IllegalArgumentException("Ehcache Disk tier cache size should be greater than " + MINIMUM_MAX_SIZE_IN_BYTES);
+        }
+        this.storagePath = Objects.requireNonNull(builder.storagePath, "Storage path shouldn't be null");
+        if (builder.threadPoolAlias == null || builder.threadPoolAlias.isBlank()) {
+            this.threadPoolAlias = THREAD_POOL_ALIAS_PREFIX + "DiskWrite";
+        } else {
+            this.threadPoolAlias = builder.threadPoolAlias;
+        }
+        this.settings = Objects.requireNonNull(builder.settings, "Settings objects shouldn't be null");
+        Objects.requireNonNull(builder.settingPrefix, "Setting prefix shouldn't be null");
+        this.DISK_WRITE_MINIMUM_THREADS = Setting.intSetting(builder.settingPrefix + ".tier.disk.ehcache.min_threads", 2, 1, 5);
+        this.DISK_WRITE_MAXIMUM_THREADS = Setting.intSetting(builder.settingPrefix + ".tier.disk.ehcache.max_threads", 2, 1, 20);
+        // Default value is 1 within EhCache.
+        this.DISK_WRITE_CONCURRENCY = Setting.intSetting(builder.settingPrefix + ".tier.disk.ehcache.concurrency", 2, 1, 3);
+        // Default value is 16 within Ehcache.
+        this.DISK_SEGMENTS = Setting.intSetting(builder.settingPrefix + "tier.disk.ehcache.segments", 16, 1, 32);
```

Comment:
Does is make sense to create constants for these values to make sure its not lost in the code and we can tune default values easily in the future?

Reply:
Yeah created constants for these now.